Source/Core/Core/GeckoCodeConfig.cpp

```diff
@@ -2,17 +2,173 @@
 // Licensed under GPLv2+
 // Refer to the license.txt file included.
 
+#include "Core/GeckoCodeConfig.h"
+
 #include <algorithm>
+#include <cstddef>
+#include <memory>
 #include <sstream>
-#include <string>
-#include <vector>
+
+#include <curl/curl.h>
 
 #include "Common/IniFile.h"
+#include "Common/Logging/Log.h"
 #include "Common/StringUtil.h"
-#include "Core/GeckoCodeConfig.h"
 
 namespace Gecko
 {
+static std::string MakeEndpointURL(const std::string& game_id)
+{
+  constexpr const char* url_prefix = "http://geckocodes.org/txt.php?txt=";
+
+  switch (game_id[0])
+  {
+  case 'R':
+  case 'S':
+  case 'G':
+    return url_prefix + game_id;
+  default:
+    // All channels (WiiWare, VirtualConsole, etc) are identified by their first four characters
+    return url_prefix + game_id.substr(0, 4);
+  }
+}
+
+static size_t DownloadCodesWriteCallback(void* contents, size_t size, size_t nmemb,
```

Comment:
you'd just `static_cast` the pointer to `std::string*` in this case.

Casting the function would be undefined behavior (as per `[expr.reinterpret_cast]/6` in the C++14 standard)

> A function pointer can be explicitly converted to a function pointer of a different type. The effect of calling a function through a pointer to a function type (8.3.5) that is not the same as the type used in the definition of the function is undefined.